File: AdForest/app/src/main/java/com/scriptsbundle/adforest/messages/adapter/ChatAdapter.java
package com.scriptsbundle.adforest.messages.adapter;


import android.app.Activity;
import android.app.DownloadManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.media.Image;
import android.net.Uri;
import android.os.Environment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.androidnetworking.AndroidNetworking;
import com.androidnetworking.common.Priority;
import com.bumptech.glide.Glide;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.io.File;
import java.util.ArrayList;

//import com.scriptsbundle.adforest.CollageImageView;
import io.reactivex.rxjava3.core.Observable;
import com.scriptsbundle.adforest.R;
import com.scriptsbundle.adforest.Settings.Settings;
import com.scriptsbundle.adforest.ad_detail.FragmentAdDetail;
import com.scriptsbundle.adforest.ad_detail.full_screen_image.FullScreenViewActivity;
import com.scriptsbundle.adforest.messages.ChatFragment;
import com.scriptsbundle.adforest.modelsList.ChatMessage;
import com.scriptsbundle.adforest.utills.CollageView;
import com.scriptsbundle.adforest.utills.Helpers;
import com.scriptsbundle.adforest.utills.SettingsMain;

import static android.content.Context.DOWNLOAD_SERVICE;

public class ChatAdapter extends BaseAdapter {

    private static LayoutInflater inflater = null;
    SettingsMain settingsMain;
    Context context;
    private ArrayList<ChatMessage> chatMessageList;

    public ChatAdapter(Activity activity, ArrayList<ChatMessage> list) {
        chatMessageList = list;
        context = activity;
        inflater = (LayoutInflater) activity
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        settingsMain = new SettingsMain(context);

    }

    @Override
    public int getCount() {
        return chatMessageList.size();
    }

    @Override
    public Object getItem(int position) {
        return position;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final ChatMessage message = chatMessageList.get(position);
        View vi = convertView;

        if (settingsMain.getRTL()){
            if (message.isMine()) {
                vi = inflater.inflate(R.layout.item_chat_layout_rtl, null);
            } else {
                vi = inflater.inflate(R.layout.item_chat_received_layout_rtl, null);
            }
        }else{
            if (message.isMine()) {
                vi = inflater.inflate(R.layout.item_chat_layout, null);
            } else {
                vi = inflater.inflate(R.layout.item_chat_received_layout, null);
            }
        }

        TextView tv_message = vi.findViewById(R.id.message);
        TextView tv_date = vi.findViewById(R.id.tv_date);
        ImageView imageView = vi.findViewById(R.id.profile_image);
        LinearLayout bubbleLayout = vi.findViewById(R.id.chat_bubble);
        LinearLayout imagesLayout = vi.findViewById(R.id.imageLayout);
        LinearLayout fileLayout = vi.findViewById(R.id.filesLayout);
        LinearLayout collageImageView = vi.findViewById(R.id.imageInner);


        Picasso.get().load(message.getImage())
                .error(R.drawable.placeholder)
                .placeholder(R.drawable.placeholder)
                .into(imageView);
        tv_message.setText(message.getBody());
        tv_date.setText(message.getDate());

        if (message.getBody() == null || message.getBody().equals("")) {
            bubbleLayout.setVisibility(View.GONE);
        }

        if (message.getImages() != null && message.getImages().size()!=0) {
            if (message.getImages().size()==1){
                View v = LayoutInflater.from(context).inflate(R.layout.collage_one,null);
                ImageView imageView1 = v.findViewById(R.id.image1);
                Glide.with(context).load(message.getImages().get(0)).into(imageView1);
                collageImageView.addView(v);
                v.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Intent i = new Intent(context, FullScreenViewActivity.class);
                        i.putExtra("imageUrls", message.getImages());
                        i.putExtra("position", 0);
                        context.startActivity(i);
                    }
                });
            }else if (message.getImages().size()==2){
                View v = LayoutInflater.from(context).inflate(R.layout.collage_two,null);
                ImageView imageView1 = v.findViewById(R.id.image1);
                Glide.with(context).load(message.getImages().get(0)).into(imageView1);
                ImageView imageView2 = v.findViewById(R.id.image2);
                Glide.with(context).load(message.getImages().get(1)).into(imageView2);
                collageImageView.addView(v);
                v.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Intent i = new Intent(context, FullScreenViewActivity.class);
                        i.putExtra("imageUrls", message.getImages());
                        i.putExtra("position", 0);
                        context.startActivity(i);
                    }
                });
            }else if (message.getImages().size()==3){
                try{

                    View v = LayoutInflater.from(context).inflate(R.layout.collage_three,null);
                    ImageView imageView1 = v.findViewById(R.id.image1);
                    Glide.with(context).load(message.getImages().get(0)).into(imageView1);
                    ImageView imageView2 = v.findViewById(R.id.image2);
                    Glide.with(context).load(message.getImages().get(1)).into(imageView2);
                    ImageView imageView3 = v.findViewById(R.id.image3);
                    Glide.with(context).load(message.getImages().get(2)).into(imageView3);
                    collageImageView.addView(v);
                    v.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            Intent i = new Intent(context, FullScreenViewActivity.class);
                            i.putExtra("imageUrls", message.getImages());
                            i.putExtra("position", 0);
                            context.startActivity(i);
                        }
                    });
                }catch (Exception e){
                    e.printStackTrace();
                }
            }else {
                View v = LayoutInflater.from(context).inflate(R.layout.collage_four,null);
                ImageView imageView1 = v.findViewById(R.id.image1);
                Glide.with(context).load(message.getImages().get(0)).into(imageView1);
                ImageView imageView2 = v.findViewById(R.id.image2);
                Glide.with(context).load(message.getImages().get(1)).into(imageView2);
                ImageView imageView3 = v.findViewById(R.id.image3);
                Glide.with(context).load(message.getImages().get(2)).into(imageView3);
                ImageView imageView4 = v.findViewById(R.id.image4);
                Glide.with(context).load(message.getImages().get(3)).into(imageView4);
                collageImageView.addView(v);
                v.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Intent i = new Intent(context, FullScreenViewActivity.class);
                        i.putExtra("imageUrls", message.getImages());
                        i.putExtra("position", 0);
                        context.startActivity(i);
                    }
                });
            }
        } else {
            imagesLayout.setVisibility(View.GONE);
        }

        if (message.getFile() != null) {
            String urlStr = message.getFile().get(0);
            String fileName = urlStr.substring(urlStr.lastIndexOf('/')+1, urlStr.length());
            TextView fileNameTv = vi.findViewById(R.id.fileName);
            fileNameTv.setText(fileName);
            fileLayout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Helpers.Companion.downloadFile(context,message.getFile().get(0),fileName);
                }
            });
        } else {
            fileLayout.setVisibility(View.GONE);
        }


        // if message is mine then align to right
        if (settingsMain.getRTL()) {
            if (message.isMine()) {
                tv_message.setPaddingRelative(10, 5, 30, 0);
                tv_date.setPaddingRelative(10, 10, 30, 0);
                bubbleLayout.setBackgroundResource(R.drawable.ic_rtl_send_message);

            }
            // If not mine then align to left
            else {
                bubbleLayout.setBackgroundResource(R.drawable.ic_rtl_received_message);
                tv_message.setPaddingRelative(30, 5, 10, 0);
                tv_date.setPaddingRelative(30, 10, 10, 0);
            }
        }

        return vi;

    }




    public void add(ChatMessage object) {
        chatMessageList.add(object);
    }
}

File: AdForest/app/src/main/java/com/scriptsbundle/adforest/messages/WhizzChat/Adapters/ChatAdapter.kt
package com.scriptsbundle.adforest.messages.WhizzChat.Adapters

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.bumptech.glide.Glide
import com.google.gson.Gson
import com.google.gson.JsonObject
import com.squareup.picasso.Picasso
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import okhttp3.*
import com.scriptsbundle.adforest.R
import com.scriptsbundle.adforest.ad_detail.full_screen_image.FullScreenViewActivity
import com.scriptsbundle.adforest.messages.WhizzChat.Models.ChatScreenModels.ChatItem
import com.scriptsbundle.adforest.utills.Helpers.Companion.downloadFile
import com.scriptsbundle.adforest.utills.SettingsMain
import java.io.IOException


class ChatAdapter(var list: List<ChatItem>, val context: Context) : BaseAdapter() {
    override fun getCount(): Int {
        return list.size
    }

    override fun getItem(p0: Int): Any {
        TODO("Not yet implemented")
    }

    override fun getItemId(p0: Int): Long {
        return p0.toLong()
    }

    val settingsMain = SettingsMain(context)
    override fun getView(position: Int, p1: View?, p2: ViewGroup?): View {
        lateinit var vi: View
        var chat = list[position]



        if (settingsMain.getRTL()){

            if (chat.is_reply == "message-sender-box") {
                vi = LayoutInflater.from(context).inflate(R.layout.whizzchat_item_chat_rtl, null)
            } else {
                vi = LayoutInflater.from(context).inflate(R.layout.whizzchat_item_chat_received_rtl, null)
            }
        }else{

            if (chat.is_reply == "message-sender-box") {
                vi = LayoutInflater.from(context).inflate(R.layout.whizzchat_item_chat_layout, null)
            } else {
                vi = LayoutInflater.from(context).inflate(R.layout.whizzchat_item_chat_received_layout, null)
            }
        }


        val tv_message = vi.findViewById<TextView>(R.id.message)
        val tv_date = vi.findViewById<TextView>(R.id.tv_date)
        val fileDate = vi.findViewById<TextView>(R.id.fileDate)
        val mapDate = vi.findViewById<TextView>(R.id.mapDate)
        val imageDate = vi.findViewById<TextView>(R.id.imageDate)
        val imageView = vi.findViewById<ImageView>(R.id.profile_image)
        val bubbleLayout = vi.findViewById<LinearLayout>(R.id.chat_bubble)
        val imagesLayout = vi.findViewById<LinearLayout>(R.id.imageLayout)
        val fileLayout = vi.findViewById<LinearLayout>(R.id.filesLayout)
        val mapView = vi.findViewById<ImageView>(R.id.map_view)
        val mapLayout = vi.findViewById<LinearLayout>(R.id.mapLayout)
        val collageImageView = vi.findViewById<LinearLayout>(R.id.imageInner)

        tv_date.text = chat.time_chat
        fileDate.text = chat.time_chat
        imageDate.text = chat.time_chat
        mapDate.text = chat.time_chat
        if (chat.message_type=="text"){
            bubbleLayout.visibility = View.VISIBLE
            imagesLayout.visibility = View.GONE
            mapLayout.visibility = View.GONE
            fileLayout.visibility = View.GONE
            tv_message.text = chat.msg
        }else if (chat.message_type=="image"){

            bubbleLayout.visibility = View.GONE
            imagesLayout.visibility = View.VISIBLE
            fileLayout.visibility = View.GONE
            mapLayout.visibility = View.GONE
            when (chat.image_url.size) {
                1 -> {
                    val v = LayoutInflater.from(context).inflate(R.layout.collage_one, null)
                    val imageView1 = v.findViewById<ImageView>(R.id.image1)
                    loadImage(chat.image_url.get(0), imageView1)
                    collageImageView.addView(v)
                    v.setOnClickListener {
                        val i = Intent(context, FullScreenViewActivity::class.java)
                        i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        i.putExtra("imageUrls", chat.image_url)
                        i.putExtra("position", 0)
                        context.startActivity(i)
                    }
                }
                2 -> {
                    val v = LayoutInflater.from(context).inflate(R.layout.collage_two, null)
                    val imageView1 = v.findViewById<ImageView>(R.id.image1)
                    loadImage(chat.image_url.get(0), imageView1)
                    val imageView2 = v.findViewById<ImageView>(R.id.image2)
                    loadImage(chat.image_url.get(1), imageView2)
                    collageImageView.addView(v)
                    v.setOnClickListener {
                        val i = Intent(context, FullScreenViewActivity::class.java)
                        i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        i.putExtra("imageUrls", chat.image_url)
                        i.putExtra("position", 0)
                        context.startActivity(i)
                    }
                }
                3 -> {
                    try {
                        val v = LayoutInflater.from(context).inflate(R.layout.collage_three, null)
                        val imageView1 = v.findViewById<ImageView>(R.id.image1)
                        loadImage(chat.image_url.get(0), imageView1)
                        val imageView2 = v.findViewById<ImageView>(R.id.image2)
                        loadImage(chat.image_url.get(1), imageView2)
                        val imageView3 = v.findViewById<ImageView>(R.id.image3)
                        loadImage(chat.image_url.get(2), imageView3)
                        collageImageView.addView(v)
                        v.setOnClickListener {
                            val i = Intent(context, FullScreenViewActivity::class.java)
                            i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                            i.putExtra("imageUrls", chat.image_url)
                            i.putExtra("position", 0)
                            context.startActivity(i)
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
                else -> {
                    val v = LayoutInflater.from(context).inflate(R.layout.collage_four, null)
                    val imageView1 = v.findViewById<ImageView>(R.id.image1)
                    loadImage(chat.image_url.get(0), imageView1)
                    val imageView2 = v.findViewById<ImageView>(R.id.image2)
                    loadImage(chat.image_url.get(1), imageView2)
                    val imageView3 = v.findViewById<ImageView>(R.id.image3)
                    loadImage(chat.image_url.get(2), imageView3)
                    val imageView4 = v.findViewById<ImageView>(R.id.image4)
                    loadImage(chat.image_url.get(3), imageView4)
                    collageImageView.addView(v)
                    v.setOnClickListener {
                        val i = Intent(context, FullScreenViewActivity::class.java)
                        i.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        i.putExtra("imageUrls", chat.image_url)
                        i.putExtra("position", 0)
                        context.startActivity(i)
                    }
                }
            }

        }else if (chat.message_type == "file"){
            bubbleLayout.visibility = View.GONE
            imagesLayout.visibility = View.GONE
            fileLayout.visibility = View.VISIBLE
            mapLayout.visibility = View.GONE

            chat.file_url.let {
                val urlStr: String = it.get(0)
                val fileName = urlStr.substring(urlStr.lastIndexOf('/') + 1, urlStr.length)
                val fileNameTv = vi.findViewById<TextView>(R.id.fileName)
                fileNameTv.text = fileName
                fileLayout.setOnClickListener { downloadFile(context, chat.file_url.get(0), fileName) }

            }
        }else if(chat.message_type == "map"){
            bubbleLayout.visibility = View.GONE
            imagesLayout.visibility = View.GONE
            fileLayout.visibility = View.GONE
            mapLayout.visibility = View.VISIBLE

            val convertedObject: JsonObject = Gson().fromJson(chat.attachments, JsonObject::class.java)
            val appInfo = context.packageManager.getApplicationInfo(context.packageName,PackageManager.GET_META_DATA)
            val apiKey = appInfo.metaData.getString("com.google.android.geo.API_KEY")
            Glide.with(context).load("http://maps.googleapis.com/maps/api/staticmap?zoom=16&size=300x160&markers=size:mid%7Ccolor:red|"+
                    convertedObject.get("latitude")+","+convertedObject.get("longitude")+"&key=$apiKey").into(mapView)
            mapView.setOnClickListener{
                val gmmIntentUri = Uri.parse("geo:0,0?q="+convertedObject.get("latitude")+","+convertedObject.get("longitude"))
                val mapIntent = Intent(Intent.ACTION_VIEW, gmmIntentUri)
                mapIntent.setPackage("com.google.android.apps.maps")
                mapIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                context.startActivity(mapIntent)
            }

        }

        return vi
    }

    private fun loadImage(url: String, imageView: ImageView){
        CoroutineScope(Dispatchers.Main).launch {
            Picasso.get().load(url).into(imageView)
        }
    }
}